Feature/paas location

kiuby88 opened the pull request:
Adding `CloudFoundryPaasLocation`.
- `CloudFoundryPaasLocation` allows to Cloud Foundry-based platforms to be pointed from Brooklyn.
- A resolver for `CloudFoundryPaasLocation` was added. Then, this location can be resolved using the  prefix `cloudfoundry`.
- Some simple unit and live tests were added.

kiuby88 committed:
Adding CloudFoundryPaasLocation.

This location allows to CloudFoundry-based platforms to be reached.

kiuby88 committed:
CloudFoundryPaasLocationResolver

A Brooklyn Resolver was added for CloudFoundryPaasLocation whose prefix of the resolver is cloudfoundry

kiuby88 committed:
Live tests for CloudFoundryPaasLocation.

A simple Live Test was added.

kiuby88 commented:
@andreaturli, would you like to review this PR?

andreaturli commented on pom.xml:
```diff
@@ -48,35 +47,53 @@
+        </dependency>
+    </dependencies>
 
     <build>
```
you don't need that

kiuby88 replied:
Then, should I delete the rat plugin?

andreaturli replied:
it will be inherit by the downstream parent project

kiuby88 replied:
@andreaturli thanks for your reply. 
I have deleted the `rat` plugin configuration and I have deleted the licenses of some classes but I did not found any the typical exception `Too many files with unapproved license`  during the project building (`mvn clean install`). I found the `brooklyn-downstream-parent` pom file (v0.10.0), and I found the `rat` plugin configuration inside `pluginManagement` element . I saw that `brooklyn-downstream-parent` uses `brooklyn-server` pom file as parent. However, I did not find any configuration for the `rat` plugin inside this pom. So, it looks that rat plugin is not being inherit. 
But, I probably lost some detail about the `rat` configuration and my thoughts are wrong.

andreaturli replied:
I think you are right

you should be able to add only:

```
   <build>
      <pluginManagement>
        <plugins>
          <plugin>
            <groupId>org.apache.rat</groupId>
            <artifactId>apache-rat-plugin</artifactId>
            <configuration>
              <excludes combine.children="append">
                <!--
                    Configuration artifacts (for installations) are based on templated defaults for 
                    the given components. These are files "without any degree of creativity" from the
                    perspective of the Brooklyn/Apache contribution.
                -->
              </excludes>
            </configuration>
          </plugin>
        </plugins>
      </pluginManagement>
    </build>
```

kiuby88 replied:
@andreaturli, Thanks for the reply.
I have simplified the rat configuration.

kiuby88 replied:
I have remove the version of the plugin because, as you have mentioned, it is not necessay

andreaturli commented on pom.xml:
```diff
@@ -32,13 +32,12 @@
-        <project.entry>org.apache.brooklyn.cloudfoundry.BrooklynCloudFoundryMain</project.entry>
-        <brooklyn.version>0.10.0-SNAPSHOT</brooklyn.version> <!-- BROOKLYN_VERSION -->
+        <brooklyn.version>0.10.0-SNAPSHOT</brooklyn.version>
+        <cloudfoundry.client.version>1.1.2</cloudfoundry.client.version>
```
why this old version?

kiuby88 replied:
Sorry, I am accustomed to use this version. Robert also pointed to this some time ago. 
Updating to 1.1.3

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+
+    @SetFromFlag("cloudfoundry.instances")
+    ConfigKey<Integer> REQUIRED_INSTANCES = ConfigKeys.newIntegerConfigKey(
+            "cloudfoundry.profile.instances", "Required instances to deploy the application", 1);
```
"Number of instances of the application" ?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+
+    @SetFromFlag("cloudfoundry.instances")
+    ConfigKey<Integer> REQUIRED_MEMORY = ConfigKeys.newIntegerConfigKey(
+            "cloudfoundry.profile.memory", "Required memory to deploy the application (MB)", 512);
```
"Memory allocated for the application" ?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationConfig.java:
```diff
@@ -0,0 +1,43 @@
+
+    @SetFromFlag("cloudfoundry.instances")
+    ConfigKey<Integer> REQUIRED_DISK = ConfigKeys.newIntegerConfigKey(
+            "cloudfoundry.profile.disk", "Required disk to deploy the application (MB)", 1024);
```
"Disk size allocated for the application" ?

kiuby88 replied:
Thanks for these three suggestions regarding the descriptions.

kiuby88 committed:
Minor fixes.

Updating cloud foundry client version.
Rephrasing some ConfigKey descriptions.

kiuby88 commented:
@andreaturli thanks for the review. Just a comment form my side.

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,88 @@
+    public static ConfigKey<String> CF_ENDPOINT = ConfigKeys.newStringConfigKey("endpoint");
+    public static ConfigKey<String> CF_SPACE = ConfigKeys.newStringConfigKey("space");
+
+    CloudFoundryClient client;
```
private maybe?

kiuby88 replied:
totally agree

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,88 @@
+        return client;
+    }
+
+    public void setClient(CloudFoundryClient client) {
```
do you really need to change it ? I think setClient is not needed

kiuby88 replied:
I think `setClient` it would be very useful for the `CloudFoundryPaasLocation` testing. This method allows to mock a client and sent to the location.

andreaturli replied:
why not simply use the constructor then?

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public CloudFoundryClient getClient() {
+        return client;
```
maybe `getClient` does the setup and return the client if not null

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public void setUpClient() {
+        if (client == null) {
```
I think you can move this logic into `getClient`
Be sure it is thread-safe

kiuby88 replied:
Good idea. Let me try

kiuby88 committed:
Some refactorizations.

client is a private field inside CloudFoundryPaasLocation
setUp is marked as synchronized and as private.
getClient has to setUp the client.

Some tests were updated according to the modifications that were described above.

kiuby88 commented:
@andreaturli thanks for your last review. I have added a couple of comments.

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,89 @@
+
+    private CloudFoundryClient client;
+
+    public CloudFoundryPaasLocation() {
```
change it to `public CloudFoundryPaasLocation(CloudFoundryClient client)` and remove `setUpClient` and `setClient`

kiuby88 replied:
Thanks, just a question. 

This allows to create a location easily,  it should be very useful for unit testing.

```
    public CloudFoundryPaasLocation(CloudFoundryClient client) {
        super();
        this.client = client;
    }
```

However, I think the client generation should be `CloudFoundryPaasLocation` responsability.  For example, we would create the location through `LocationRegistry`:

```
     Map<String, String> m = MutableMap.of();
     m.put("user", "super_user");
     m.put("password", "super_secret");
     m.put("org", "secret_organization");
     m.put("endpoint", "https://api.super.secret.io");
     m.put("space", "development");

    return (CloudFoundryPaasLocation)
        mgmt.getLocationRegistry().getLocationManaged("cloudfoundry", m);
```

Then, `LocationRegistry` can not create a `client` and call `new CloudFoundryPaasLocation(client);`. So, it looks that `CloudFoundryPaasLocation` should create the client in this case. Then, I would suggest something like the following:

```
    public CloudFoundryPaasLocation(){
        super();
        client =new CloudFoundryClient(
                       new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
                       getTargetURL(getConfig(CF_ENDPOINT)), getConfig(CF_ORG), 
                       getConfig(CF_SPACE), true);
   }
```

However, the previous constructor will not be able to create the correct client because in that moment the `config keys` are empty (default values).

Of course, something like the following constructor will create the client correctly, but I think we need to ensure that a location can be created correctly using an empty constructor.

```
    public CloudFoundryPaasLocation(Map<?, ?> properties) {
        super(properties);
        client =new CloudFoundryClient(
                       new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
                       getTargetURL(getConfig(CF_ENDPOINT)), getConfig(CF_ORG), 
                       getConfig(CF_SPACE), true);
        client.login(); //login could be calling from getClient(), lazy login.
    }
```

WDYT?

andreaturli replied:
you are getting there, look at `JcloudsLocation` for the final answer :P

kiuby88 replied:
could we setup the `client` from `init`?

```
    public CloudFoundryPaasLocation() {
        super();
    }

    public CloudFoundryPaasLocation(Map<?, ?> properties) {
        super(properties);
    }

    public CloudFoundryPaasLocation(CloudFoundryClient client) {
        super();
        this.client = client;
    }

    @Override
    public void init() {
        super.init();
        if (client == null) {
            client = new CloudFoundryClient(new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
                    getTargetURL(getConfig(CF_ENDPOINT)), getConfig(CF_ORG), getConfig(CF_SPACE), true);
            client.login();
        }
    }
```

andreaturli replied:
I think so. I'd remove the  `public CloudFoundryPaasLocation(CloudFoundryClient client)` though, and add `getCloudFoundryClient`

kiuby88 replied:
Thanks for letting me find the answer :D
I agree about `getCloudFouncyClient` method.
I think `public CloudFoundryPaasLocation(CloudFoundryClient client)` it will be useful in order to test the location. For example, 

```
    @Mock
    CloudFoundryClient client;

    @Test
    public void testClientSingletonManagement() {
        cloudFoundryPaasLocation = new CloudFoundryPaasLocation(client);
        CloudFoundryClient client1 = cloudFoundryPaasLocation.getClient();
        CloudFoundryClient client2 = cloudFoundryPaasLocation.getClient();
        assertNotNull(client1);
        assertEquals(client1, client2);
    }
```

WDYT?

andreaturli replied:
is it that test really useful?
Again, I don't think we need the constructor, [KISS](https://en.wikipedia.org/wiki/KISS_principle)!

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,89 @@
+    public static ConfigKey<String> CF_ENDPOINT = ConfigKeys.newStringConfigKey("endpoint");
+    public static ConfigKey<String> CF_SPACE = ConfigKeys.newStringConfigKey("space");
+
+    private CloudFoundryClient client;
```
make it `final`

andreaturli commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,89 @@
+        }
+    }
+
+    protected CloudFoundryClient getClient() {
```
make it `public` so that you just return `client`

kiuby88 replied:
I agree, but I think an external class should not use this method

andreaturli commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationTest.java:
```diff
@@ -0,0 +1,75 @@
+    @BeforeMethod
+    public void setUp() throws Exception {
+        super.setUp();
+        MockitoAnnotations.initMocks(this);
```
does it work with `testng`?

kiuby88 replied:
Yes, I have used and I did not found any issue.

kiuby88 committed:
Moving CloudFoundryClient to CloudFoundryPaasLocation constructor.

This movement limits the mutability of the location, whether is very useful.

kiuby88 commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,88 @@
+        if (client == null) {
+            client = new CloudFoundryClient(new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
+                    getTargetURL(getConfig(CF_ENDPOINT)), getConfig(CF_ORG), getConfig(CF_SPACE), true);
+            client.login();
```
@andreaturli . I think login should not be done from `init`. Two different reasons:
- When Brooklyn builds the catalog and build an location instance, the client will login in the platform always, although the location will not being used.
- It complicates the testing, for example for testing the resolver (where we use fake data) the client tries to login in the platform and it will produce an error. Please, take a look to [this test](https://github.com/kiuby88/brooklyn-cloudfoundry/blob/feature/paas-location/src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationResolverTest.java#L73). It will produces an error because it tries to login using fake data. 

Probably, login should be offered as a location operation or the login instruction should be included in the `getClient` method. (It should be thread-safe).
WDYT?

andreaturli replied:
agreed, init() doens't need to call login on the client

kiuby88 committed:
Moving login oto getClient method.

The client does login if the accessToken is null or if the accessToken is expired.

kiuby88 commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    public CloudFoundryClient getClient() {
+        login();
```
@andreaturli `login()` method is  called from `getClient`. `login()` is thread-safe, you can find `login()` declaration above. 
(It looks that we will need to wrap the `client` in another class).

andreaturli replied:
why you need to do that inside `getClient`?

it is just a getter ...

kiuby88 replied:
Because I thought that it will ensure the token it will not expire. But, probably it is not the best way to ensure it. So, better as an independent method.

kiuby88 committed:
Moving login
Removing CloudFoundryPaasLocation constructor that uses a CloudFoundryClient as a input patameter

Deleting conctructor

kiuby88 commented on src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocation.java:
```diff
@@ -55,11 +55,6 @@ public CloudFoundryPaasLocation(Map<?, ?> properties) {
         super(properties);
     }
 
-    public CloudFoundryPaasLocation(CloudFoundryClient client) {
```
deleted!!

kiuby88 commented on src/test/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasLocationTest.java:
```diff
@@ -39,25 +41,24 @@
-    private CloudFoundryPaasLocation createCloudFoundryPaasLocation(CloudFoundryClient client) {
-        return new CloudFoundryPaasLocation(client);
+        return (CloudFoundryPaasLocation)
+                mgmt.getLocationRegistry().getLocationManaged("cloudfoundry", m);
```
produces an error, because when the client is `SetUp` in init, the endpoint `https://api.super.secret.io` is not reachable:

```
new CloudFoundryClient(
    new CloudCredentials(getConfig(CF_USER), getConfig(CF_PASSWORD)),
    getTargetURL(getConfig(CF_ENDPOINT)), //requires an reachable and existent CF platform
    getConfig(CF_ORG),
    getConfig(CF_SPACE), true);
```

I do not think we should use an existent endpoint (such as `https://api.run.pivotal.io` ) because it is an unit test and the tests should be independent of the external environment. Then, I think we need to mock the `client`. I do not think we have to mock the CF platform (for example using mockito server) in order to simulate an reachable CF platform.

WDYT?

kiuby88 commented:
I am going to apply the new design to `PaasLocation`
